lib/logic/player/favorites.dart

```diff
@@ -1,27 +1,133 @@
+import 'package:firebase_crashlytics/firebase_crashlytics.dart';
+import 'package:flutter/foundation.dart';
 import 'package:sweyer/sweyer.dart';
 
-// class FavoriteRepository {
-//   FavoriteRepository._();
-//   static final instance = FavoriteRepository._();
+@visibleForTesting
+class FavoritesRepository {
+  final serializersMap = ContentMap<IntSerializerType>({
+    for (final contentType in Content.enumerate())
+      contentType: IntListSerializer('favorites_${ContentUtils.contentTypeString(contentType)}.json')
+  });
+}
 
-//   final _serializers = ContentMap<IntListSerializer>();
+class FavoritesControl extends Control {
+  static FavoritesControl instance = FavoritesControl();
 
-//   @override
-//   Future<void> markFavorite<T extends Content>(bool value) async {
-//     assert(T != Content);
-//     _serializers.getValue<T>();
-//   }
-// }
+  @visibleForTesting
+  final repository = FavoritesRepository();
 
-// class _FavoriteBatch {
-//   bool committed = false;
+  final _favoriteSetsMap = ContentMap<Set<int>>();
 
-//   void commit() {
-//     assert(!committed);
-//     committed = true;
-//   }
+  bool _useAndroidFavorites(Type contentType) =>
+    contentType == Song && DeviceInfoControl.instance.useScopedStorageForFileModifications;
 
-//   void markFavorite<T extends Content>(bool value) {
-//     assert(!committed);
-//   }
-// }
\ No newline at end of file
+  @override
+  Future<void> init() async {
+    super.init();
+    _favoritesModeNotifier.value = false;
+    for (final contentType in Content.enumerate()) {
+      final favoriteSet = _favoriteSetsMap.getValue(contentType) ?? {};
+      if (_useAndroidFavorites(contentType)) {
+        for (final song in ContentControl.instance.state.allSongs.songs) {
+          if (song.isFavoriteInMediaStore!) {
```

Comment:
Also, at startup, should we unfavour all songs that are not favorite in the media store? Maybe this could be an additional setting, or the the setting could be: Sync with Android media store? No - Only update - Both ways

Reply:
> I assume song.isFavoriteInMediaStore is updated when the app starts? Do we also react to changes in the media store that happen while we are running? If another app marks something as a favorite.

Good catch, I will look into what can be done here!

> Also, at startup, should we unfavour all songs that are not favorite in the media store?

At startup, all maps in _favoriteSetsMap are empty.

Reply:
Added a MediaStoreContentObserver which sync the data with MediaStore